Performance 2021 chapter markdown (#2434)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/performance.md

title: Performance
description: Performance chapter of the 2021 Web Almanac covering Core Web Vitals (Largest Contentful Paint, Cumulative Layout Shift, First Input Delay) as well as First Contentful Paint and Time to First Byte
authors: [siakaramalegos]
reviewers: [rviscomi, kevinfarrugia, estelle, ziemek-bucko, jzyang, fili, samarpanda, edmondwwchan]
analysts: [siakaramalegos, rviscomi, Nithanaroy]
editors: [jzyang]
translators: []
results: https://docs.google.com/spreadsheets/d/13xhPx6o2Nowz_3b3_5ojiF_mY3Lhs25auBKM6eqGZmo/
siakaramalegos_bio: Sia Karamalegos is a web developer, international conference speaker, and writer. She is a Google Developer Expert in Web Technologies, a Cloudinary Media Developer Expert, a Stripe Community Expert, and co-organizes the Eleventy Meetup. Check out her writing, speaking, and newsletter on <a hreflang="en" href="https://sia.codes/">sia.codes</a> or find her on <a hreflang="en" href="https://twitter.com/thegreengreek">Twitter</a>.
featured_quote: TODO
featured_stat_1: TODO
featured_stat_label_1: TODO
featured_stat_2: TODO
featured_stat_label_2: TODO
featured_stat_3: TODO
featured_stat_label_3: TODO
unedited: true
---

## Introduction

> Performance is important for the user experience. Slow-to-load and slow-to-respond websites frustrate users and cause lost conversions. This is the first year that the <a hreflang="en" href="https://web.dev/vitals/">Core Web Vitals</a> have contributed to Google search rankings. As such, we've seen greater interest in improving website performance which is great news for users.

Author: I think of Core Web Vitals as the group of 3 metrics. But both you and Estelle suggested removing the "the". I've seen it used both ways. I'll go ahead and remove but I'll need to change the verb conjugation also so noting now on paper and will resolve once I can get all the other edits merged.

<i>What are our top takeaways from this year's report?</i> First, we still have a long way to go in providing a good user experience. For example, faster networks and devices have not yet reached the point where we can ignore how much JavaScript we deliver to a site; and, we may never get there. Second, sometimes we misuse new features for performance, resulting in poorer performance. Third, we need better metrics for measuring interactivity, and those are on the way. And fourth, CMS- and framework-level work on performance can significantly impact user experience for the top 10M websites.

<i>What's new this year?</i> We're excited to share performance data by traffic ranking for the first time. We also have all the core performance metrics from previous years. Finally, we added a deeper dive into the Largest Contentful Paint (LCP) element.

### Notes on Methodology

> One thing that makes the performance chapter different from the others is that we rely heavily on the <a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a> (CrUX) for our analyses. Why? If our number one priority is user experience, then the best way to measure performance is with real user data (real user metrics, or RUM for short).

Author: For the first edit - I think of it as present continuous because it's always the case. Previous years, this year, and future years will rely heavily on it.

For questions-answer style edit... My personal style is to ask questions I think the reader is asking themselves. Kind of like this - https://bloggingwizard.com/improve-your-writing-style/

Just a way to make reading it more engaging.

> One thing that makes the performance chapter different from the others is that we rely heavily on the <a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a> (CrUX) for our analyses. Why? If our number one priority is user experience, then the best way to measure performance is with real user data (real user metrics, or RUM for short).

Reviewer: OK.  For me the rhythm of the read was off because you don't generally have 1 word sentences in the majority of your writing.  This makes sense because this is not a blog; a lot of what you're presenting is based on numbers and facts and your experience in the industry.  Which was why I was saying that it has a more professional writing style for the majority.  It was off-putting to be jumping between these different rhythms in your writing: Between the formal and the super casual.  At least, that was the case for me.

> One thing that makes the performance chapter different from the others is that we rely heavily on the <a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a> (CrUX) for our analyses. Why? If our number one priority is user experience, then the best way to measure performance is with real user data (real user metrics, or RUM for short).

Reviewer: But, again, you got to do what feels right to you.  :) Just go with what makes you feel good, I'm only making suggestions.

> The Chrome User Experience Report provides user experience metrics for how real-world Chrome users experience popular destinations on the web. -<a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a>

> CrUX data only provides high-level field/RUM metrics and only for the Chrome browser. Additionally, CrUX reports data by origin, or website, instead of by page.

Author: The context here is that "website" is a definition of origin, not another alternate case.

  caption="Good Core Web Vitals by Device from 2020 to 2021",
  description="Bar chart showing the percent of origins with good Core Web Vitals in 2020 and 2021. In 2020, 34% of desktop websites were good and 24% of mobile websites. In 2021, 41% of desktop websites were good, and 29% of mobile ones.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=116267418&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

<!-- TODO: note style -->
<p class="note"><strong>Note:</strong> As the CLS calculation changed since last year, this is not an apples-to-apples comparison.</p>

The Core Web Vitals for websites in the Chrome User Experience Report improved year-over-year. But, a good part of this improvement could be due to a change in the CLS calculation, not necessarily to a performance improvement in CLS. The resulting CLS "improvement" was 8 points on desktop (2 for mobile). LCP improved by 7 points for desktop (2 for mobile). FID was already at 100% for desktop for both years and improved by 10 points on mobile.

> As in previous years, performance was better on desktop machines than mobile devices. This is why it's crucial to test your site's performance on real mobile devices and to measure real user metrics (i.e., field data). Emulating mobile in developer tools is convenient in the lab but not representative of real user experiences.

Author: The comma is needed for correct usage in English https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/.

I used "lab" to reinforce the meaning of "lab metrics" but will clarify that also means during development.

> As in previous years, performance was better on desktop machines than mobile devices. This is why it's crucial to test your site's performance on real mobile devices and to measure real user metrics (i.e., field data). Emulating mobile in developer tools is convenient in the lab but not representative of real user experiences.

Reviewer: Oh!  I learned something about i.e. with the comma then.  I have been misusing it all this time!  Thanks, @siakaramalegos !

{{ figure_markup(
  image="performance-FCP-by-ect.png",
  caption="FCP performance by connection type",
  description="Stacked bar chart showing FCP performance for effective connection type. For offline websites, 38% had good, 26% needs improvement, and 36% poor performance. For slow 2G, 100% were poor. For 2G, 1% were needs improvement, and 99% poor. For 3G, 4% were good, 21% needs improvement, and 75% poor. For 4G, 46% were good, 35% needs improvement, and 19% poor.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=679577784&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_eff_connection_type.sql"
  )
}}

Origins at 3G speeds and below experienced significant degradations in FCP. Again, ensure that you are profiling your website using real devices and networks that reflect your user data from analytics. Your JavaScript bundles may not seem significant when you're only profiling on high-end desktops with fiber connections.

> Offline connections were closer in performance to 4G though not quite as good. Service worker start-up time plus multiple cache reads could be some causes. More factors come into play with FCP than with TTFB.

Reviewer: > plus multiple cache reads could be some causes.

to

> plus multiple cache reads could have contributed.

{{ figure_markup(
  image="performance-FCP-by-rank.png",
  caption="FCP performance by rank",

>   description="Stacked bar chart showing FCP performance for each ranking group. For the top 1,000 websites, 67% had good, 28% needs improvement. For the top 10,000, 62% were good, 31% needs improvement, and 7% poor. For the top 100,000, 54% were good, 35% needs improvement, and 12% poor. For the top 1,000,000, 49% were good, 35% needs improvement, and 16% poor. For all, 45% were good, 35% needs improvement, and 20% poor.",

Reviewer: > 7% poor

to 

> 7% were poor

...

> 12% were poor

...

> 16% were poor

...

> 20% were poor

Trying to keep your lists consistent.  There's always the `were good` earlier and then the inconsistent `poor` later.

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=965739974&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

> Like TTFB, FCP improved with higher rankings. Also like TTFB, only <a hreflang="en" href="https://datastudio.google.com/s/kZ9K0d-sBQw">19.5% of origins on WordPress experienced good FCP performance</a>. Since their TTFB performance was poor, I am not surprised that their FCP is also slow. It's difficult to achieve good scores on FCP and LCP if TTFB is slow.

Reviewer: Again, this is a business/research/professional work.  Try to remove the first person and use the third person instead (a.k.a. avoid writing that makes it look like a personal opinion, which makes it seem less factual and less reliable).  I haven't touched a lot of your "I"s, but that's the general rule for this type of writing.  Up to you if you want to take this advice.

Common culprits for poor FCP are render-blocking resources, server response times (anything associated with a slow TTFB), large network payloads, and more.

### Largest Contentful Paint (LCP)

<a hreflang="en" href="https://web.dev/lcp/">Largest Contentful Paint (LCP)</a> measures the time from start load to when the browser renders the largest image or text in the viewport.

{{ figure_markup(
  image="performance-LCP-by-device.png",
  caption="LCP performance by device",

>   description="Stacked bar chart showing LCP performance for desktop and mobile devices. For desktop devices, 60% of websites had good (< 2.5s), 27% needs improvement, and 12% poor (>= 4s) LCP performance. For mobile, 45% were good, 35% needs improvement, and 19% poor.",

Reviewer: `good` and `poor` do not match, i.e.:
> 12% had poor

{{ figure_markup(
  image="performance-LCP-by-ect.png",
  caption="LCP performance by connection type",
  description="Stacked bar chart showing LCP performance for effective connection type. For offline websites, 49% had good, 17% needs improvement, and 34% poor performance. For slow 2G, 1%  were needs improvement and 99% poor. For 2G, 4% were needs improvement and 96% poor. For 3G, 8% were good, 28% needs improvement, and 65% poor. For 4G, 51% were good, 33% needs improvement, and 16% poor.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=583947675&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_eff_connection_type.sql"
  )
}}

Offline origins with good LCP more closely matched 4G experiences, though poor LCP experiences were higher for offline. LCP occurs after FCP, and the additional budget of 0.7 seconds could be why more offline websites achieved good LCP than FCP.

{{ figure_markup(
  image="performance-LCP-by-rank.png",
  caption="LCP performance by rank",

>   description="Stacked bar chart showing LCP performance for each ranking group. For the top 1,000 websites, 64% had good, 25% needs improvement, and 11% poor performance. For the top 10,000, 59% were good, 30% needs improvement, and 11% poor. For the top 100,000, 55% were good, 32% needs improvement, and 13% poor. For the top 1,000,000, 53% were good, 33% needs improvement, and 14% poor. For all, 50% were good, 33% needs improvement, and 17% poor.",

Reviewer: `good` and `poor` mismatch

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1080361066&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

For LCP, the differences in performance by rank were closer than FCP. Also, a higher proportion of origins in the top 1,000 had poor LCP. On WordPress, <a hreflang="en" href="https://datastudio.google.com/s/kvq1oJ60jaQ">28% of origins experienced good LCP</a>. This is an opportunity to improve user experience as poor LCP is usually caused by a handful of problems.

### The LCP Element

> Let's take a deeper dive look at the LCP element.

Reviewer: > Let's take a deeper dive look at the LCP element.

to

> Let's take a deeper dive into the LCP element.

{{ figure_markup(
  image="performance-top-15-lcp-nodes.png",
  caption="Top 15 LCP HTML element nodes",
  description="Bar chart showing the percent share of pages for each node type for the LCP element- data listed by desktop then mobile: IMG 47% 42%, DIV 28% 27%, P 6% 10%, H1 3% 5%, SECTION 3% 2%, undetected 2% 2%, H2 2% 2%, A 1% 2%, SPAN 1% 1%, HEADER 1% 1%, LI 1% 1%, RS-SBG 1% 1%, H3 0% 1%, TD 1% 1%, VIDEO 0% 0%",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1549429021&format=interactive",
  width=600,
  height=909,
  sheets_gid="730149797",
  sql_file="lcp_element_data.sql"
  )
}}

> IMG, DIV, P, and H1 made up 83% of all LCP nodes (on mobile). This doesn't tell us if the content was an image or text though as background images can be applied with CSS.

Reviewer: > This doesn't tell us if the content was an image or text though as background images can be applied with CSS.

to

> This doesn't tell us if the content was an image or a text, as background images can be applied with CSS.

{{ figure_markup(
  image="performance-lcp-antipatterns.png",
  caption="LCP elements with performance anti-patterns",
  description="Bar chart showing 9.3% of pages on mobile use native lazy load on the LCP element, 16.5% probably lazy load, and 0.4% use async decode.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=130632749&format=interactive",
  sheets_gid="730149797",
  sql_file="lcp_element_data.sql"
  )
}}

For user experience, we want LCP elements to load as fast as possible. User experience is why LCP was selected as one of the Core Web Vitals. We do not want it to be lazy-loaded nor decoded asynchronously as that further delays the render. However, we can see that 9.3% of pages used the native loading=lazy flag on the LCP `<img>` element.

Not all browsers support native lazy loading. Popular lazy loading polyfills detect a "lazyload" class on an image element. Thus, we can identify more possibly lazy-loaded images by adding images with a "lazyload" class to the total. The percent of sites probably lazy loading their LCP `<img>` element jumps up to 16.5% on mobile.

> Lazy-loading your LCP element will result in worse performance. Don't do it! WordPress was an early adopter of native lazy loading. The early method was a naive solution applying lazy loading to all images, and the results showed a <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">negative performance correlation</a>. They were able to use this data to implement a more nuanced approach for better performance.

Reviewer: You can try to put emphasis around your: Don't do it!

Italics?

Inconsistency in how you write: Lazy-loading, as you have `lazy loading` later.  Pick one and stick to it :)

> Luckily, only 0.4% of sites used the async decode directive for their LCP image. And, the negative impact of asynchronous decode is not as high as lazy loading.

Reviewer: > And, the negative impact of asynchronous decode is not as high as lazy loading.

to

> The negative impact of asynchronous decode is not as high as lazy loading.

{{ figure_markup(
  content="354",
  caption="Websites attempted to use native lazy-loading on LCP elements that are not images or iframes",
  classes="big-number",
  sheets_gid="1423728540",
  sql_file="lcp_element_data.sql"
  )
}}

> Interestingly, 354 origins on desktop attempted to use native-lazy loading on HTML elements that do not support the loading attribute (e.g., `<div>`). The loading attribute is only supported on `<img>` and  sometimes `<iframe>` elements (see <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">caniuse</a>).

Reviewer: > native-lazy loading

to

> native lazy-loading

Author: lol

### Cumulative Layout Shift (CLS)

{{ figure_markup(
  image="performance-CLS-by-device.png",
  caption="CLS performance by device",

>   description="Stacked bar chart showing CLS performance for desktop and mobile devices. For desktop devices, 62% of websites had good (< 0.1), 23% needs improvement, and 15% poor (>= 0.25) CLS performance. For mobile, 62% were good, 21% needs improvement, and 17% poor.",

Reviewer: inconsistent between `good` and `poor`

Author: I'm confused what you mean by "inconsistent"? Are the numbers wrong?

Reviewer 2: Not sure what she means by that either, but did spot something else in this description: "good (< 0.1)....poor (>= 0.25)". all CWV borderlines were clarified as to be in the better category a while back. So my understand is good is <= 0.1 and poor is > 0.25.

https://web.dev/vitals/

![image](https://user-images.githubusercontent.com/10931297/141685951-9a300539-7cb1-4aff-b79b-d9938f59aa7c.png)

Author: @tunetheweb I think it has to do with verb usage... see later comment I made

Reviewer 2: I was just clarifying that the point I was about to make, was a completely separate point (but one that should be addressed).

Author: OH. I did not make the charts and assumed they were all correct. If that's the case then we'll need to redo them all 😭

Author: Per rick "The charts are correct when we use CrUX BQ as the data source. The underlying histograms use < not <= for the upper limit of each bin." Marking as resolved.

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=160840238&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

<a hreflang="en" href="https://web.dev/cls/">Cumulative Layout Shift (CLS)</a> is characterized by how much shift a user experiences, not how long it takes to visually see something like FCP and LCP. As such, performance by device was fairly equivalent.

{{ figure_markup(
  image="performance-CLS-by-ect.png",
  caption="CLS performance by connection type",

>   description="Stacked bar chart showing CLS performance for effective connection type. For offline websites, 87% had good and 4% poor performance. For slow 2G and 2G, 53-52% were good, 15% needs improvement, and 32-33% poor. For 3G, 58% were good, 16% needs improvement, and 26% poor. For 4G, 69% were good, 13% needs improvement, and 18% poor.",

Reviewer: inconsistent between `good` and `poor`

Reviewer: The inconsistent is about it being `had good` and just `poor` instead of `had poor`

Author: Then the inconsistency applies to needs improvement also.

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1650043738&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_eff_connection_type.sql"
  )
}}

> Performance degradation from 4G to 3G and below was not as pronounced as with FCP and LCP. Some degradation exists, but it's not reflected in the device data, only the connection type.

Reviewer: > Performance degradation from 4G to 3G and below

to

> Performance degradation from `4G` to `3G and below`

I get what you're writing here, it's just not enough visual cues if I was skimming.  Just trying to give you a clear idea of what I'm looking for.  Maybe italics would be more subtle and consistent but still clear?

Author: I think italics is for emphasis which would seem out of context here. You're just wanting clarification. Not sure I can think of a way at the moment so leaving this up.

Reviewer 2: Is it:
> "4G" to "3G" and below
Or is it:
> "4G" to "3G and below"

I presume it's the former as there are 2G buckets as well, in which case not sure it's necessary to give it a different style? If it's the latter, then I agree we should make it obvious "3G and below" is one category with either code formatting or quotes.

We generally use italics to introduce new terms rather than for emphasis. As this is not the first introduction of the 4G term so, by that, italics wouldn't be appropriate here. However, it _can_ be used for emphasis but IMHO most writers overuse italics and bold for emphasis where it's not needed, which is why we discourage its use in our wikis. But there's always exceptions, and maybe this is one?

Author: It's both? Basically 4G is faster and everything else no matter how you slice it is slower, except offline of course.

Should I remove the other italics then? Didn't realize this was in the style guide.

Reviewer 2: It's in the Editor's Guide: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Editors'-Guide

> Try to avoid over-use of bold and even italics – Call out figures should be used for important stats, and italics should be used for introducing terms.

If https://github.com/HTTPArchive/almanac.httparchive.org/commit/a8f68a3559340fc9e163ef0e7aff565e1f030d3f is all the italics you had that was fine and feel free to put back.

It's _people_ who, seemingly _randomly_ decide to use _italics_, for many _words_ mid-sentence, apparently for _emphasis_, that can be _distracting_ that we discourage. Smashing Magazine used this style (though with __bold__) but I don't particularly like it to be honest.

The number of arguments I've had with authors on this point over the years! 😀

Offline websites had the highest CLS performance of all connection types. For sites with service worker caching, some assets like images and ads that would otherwise cause layout shifts may not be cached. Thus, they would never load and never cause a layout shift. Often fallback HTML for these sites can be more basic versions of the online website.

{{ figure_markup(
  image="performance-CLS-by-rank.png",
  caption="CLS performance by rank",

>   description="Stacked bar chart showing CLS performance for each ranking group. For the top 1,000 websites, 53% had good, 25% needs improvement, and 21% poor performance. For the top 10,000, 46% were good, 27% needs improvement, and 27% poor. For the top 100,000, 48% were good, 26% needs improvement, and 26% poor. For the top 1,000,000, 54% were good, 25% needs improvement, and 21% poor. For all, 61% were good, 23% needs improvement, and 16% poor.",

Reviewer: inconsistent between `good` and `poor`

Author: Maybe you want the verb for every category? I figured that would make the screen reader time even longer. For "needs improvement" and "poor", both are bucket names so was just trying to get the numbers across quickly and only provide the verb for the first.

Author: If I'm understanding correctly, that would change the description for this and all similar to something like:

> Stacked bar chart showing CLS performance for each ranking group. For the top 1,000 websites, 53% had good, 25% had needs improvement, and 21% had poor performance. For the top 10,000, 46% were good, 27% were needs improvement, and 27% were poor. For the top 100,000, 48% were good, 26% were needs improvement, and 26% were poor. For the top 1,000,000, 54% were good, 25% were needs improvement, and 21% were poor. For all, 61% were good, 23% were needs improvement, and 16% were poor.

Reviewer: You can also remove `had`?  I just meant you had one part of the list with the verb and the other parts of the list without.  I think if I were being strictly grammatical, I'd need the verb in there?

  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

For ranking, CLS performance showed an interesting trough for the top 10,000 websites. In addition, all the ranked groups above 1M performed worse than the sites ranked under 1M. Since the "all"  group had better performance than all the other ranked groupings the sub-1M group performs better. WordPress may again play a role in this as <a hreflang="en" href="https://datastudio.google.com/s/qG00yMxSa3o">60% of origins on WordPress experienced a good CLS</a>.

Common culprits for poor CLS include not reserving space for images, text shifts when web fonts are loaded, top banners inserted after first paint, non-composited animations, and iframes.

### First Input Delay (FID)

<a hreflang="en" href="https://web.dev/fid/">First Input Delay (FID)</a> measures the time from when a user first interacts with a page to the time the browser begins processing event handlers in response to that interaction.

{{ figure_markup(
  image="performance-FID-by-device.png",
  caption="FID performance by device",

>   description="Stacked bar chart showing FID performance for desktop and mobile devices. For desktop devices, 100% of websites had good (< 100ms). For mobile, 90% were good, 10% needs improvement, and 0% poor (>= 300ms).",

Reviewer: inconsistent between `good` and `poor` but maybe you don't even need the 0%?

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=230841623&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

FID performance was better on desktop than on mobile devices likely due to device speeds which can better handle larger amounts of JavaScript.

{{ figure_markup(
  image="performance-FID-by-ect.png",
  caption="FID performance by connection type",

>   description="Stacked bar chart showing FID performance for effective connection type. For offline websites, 78% had good, 17% needs improvement, and 5% poor performance. For slow 2G and 2G, 81-82% were good, 18% needs improvement, and 0-1% poor. For 3G, 90% were good and 10% needs improvement. For 4G, 93% were good, 7% needs improvement, and 0% poor.",

Reviewer: inconsistent between `good` and `poor`

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1135701788&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_eff_connection_type.sql"
  )
}}

> FID performance degraded some by connection type, but less so than the other metrics. The high distribution of scores seems to be reducing the amount of variance we're seeing in the analysis.

Reviewer: > seems

to

> seemed

Author: That sounds weird to me in this case. Like it's still causing the lack of variance - it's not something that stopped happening in the past.

Reviewer: I suggested past tense because you're talking about a conclusion based on the data that was measured in the past and deducing what the past data looked to you.  The distribution, variance and analysis were all done in the past.  If it sounds weird to you, then keep it.

Unlike the other metrics, FID was worse for offline websites than any other connection category. This could be due to the more complex nature of many websites with service workers. Having a service worker does not eliminate the impact of client-side JavaScript running on the main thread.

{{ figure_markup(
  image="performance-FID-by-rank.png",
  caption="FID performance by rank",

>   description="Stacked bar chart showing FID performance for each ranking group. For all categories, 93-94% had good, 5-7% needs improvement, and 0% poor performance.",

Reviewer: inconsistent between `good` and `poor`

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1539782601&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

FID performance by rank was flat.

For all FID metrics, we see very large bars in the "good" category which makes it less effective unless we've truly hit peak performance. The good news is the <a hreflang="en" href="https://web.dev/better-responsiveness-metric/">Chrome team is evaluating this now</a> and would like your feedback.

If your site's performance is not in the "good" category, then you definitely have a performance problem. A common culprit for FID issues is too much long-running JavaScript. Keep your bundle sizes small and pay attention to third-party scripts.

### Total Blocking Time (TBT)

> The Total Blocking Time (TBT) metric measures the total amount of time between First Contentful Paint (FCP) and Time to Interactive (TTI) where the main thread was blocked for long enough to prevent input responsiveness. --<a hreflang="en" href="https://web.dev/tbt/">Web.dev</a> ]

> <a hreflang="en" href="https://web.dev/tbt/">Total Blocking Time (TBT)</a> is a lab-based metric that helps us debug potential interactivity issues. FID is a field-based metric, and TBT is its lab-based analog. Currently, when evaluating client websites, I reach for total blocking time TBT as another indicator of possible performance issues due to JavaScript.

Reviewer: > I reach

to

> you can look at

or rewrite without first person

Author: This was a case where I thought it was too personal-opinion of me so wanted to clarify not necessarily the Web Almanac view

Reviewer: I am just not sure the reader makes the distinction between you and the Web Almanac.  If this information was unreliable they're more likely to think it's the Web Almanac irregardless of how you voice it.  Just, again, jarring to go from data driven text to first person.

Author: See guidance here https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#general-principles-for-technical-writing

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=2135422883&format=interactive",
  sheets_gid="730149797",
  sql_file="tbt.sql"
  )
}}

<p class="note"><strong>Note:</strong> The groups in the chart are based off of the Lighthouse score for TBT (e.g., >= 0.9 results in "good"). Due to rounding of the score, some TBT values slightly above 200ms get categorized as "good" (and similarly at the 600ms threshold).</p>

Remember that the data is a single, throttled-CPU Lighthouse run through WebPageTest and does not reflect real user experiences. Yet, potential interactivity looked much worse when looking at TBT versus FID. The "real" evaluation of your interactivity is probably somewhere between. Thus, if your FID is "good", take a look at TBT in case you're missing some poor user experiences that FID can't catch yet. The same issues that cause poor FID also cause poor TBT.

{{ figure_markup(
  content="67 seconds",
  caption="Longest TBT",
  classes="big-number",
  sheets_gid="1423728540",
  sql_file="lcp_element_data.sql"
  )
}}

## Conclusion

> We still have a long way to go to provide great user experiences, but we can take steps to improve them.

Reviewer: > experiences

to

> experience

First, you cannot improve performance unless you can measure it. A good first step here is to measure your site using real user devices and to set up real-user monitoring (RUM). You can get a flavor of how your site performs with Chrome users with the <a hreflang="en" href="https://rviscomi.github.io/crux-dash-launcher/">CrUX dashboard launcher</a> (if your site is in the dataset). You should set up a RUM solution that measures across multiple browsers. You can build this yourself or use one of many analytics vendors' solutions.

Second, as new features in HTML, CSS, and JavaScript are released, make sure you understand them before implementing them. Use A/B testing to verify that adopting a new strategy results in improved performance. For example, don't lazy-load images above the fold. If you have a RUM tool implemented, you can better detect when your changes accidentally cause regressions.

> Additionally, continue to optimize for both FID (field/real-user data) and TBT (lab data). Take a look at the <a hreflang="en" href="https://web.dev/responsiveness/">proposal</a> for a new responsiveness metric and participate by providing feedback. A new <a hreflang="en" href="https://web.dev/smoothness/">animation smoothness metric</a> is also being proposed. In our quest for a faster web, change is inevitable and for the better. As we continue to optimize, you're participation is key.

Reviewer: > Additionally,

to

> Third

lol, you don't seem to like counting above two?  Just teasing.

Author: I think it can get confusing as in the reader has probably forgotten the other numbers. :) Will change it though.

> Finally, we saw that WordPress can impact the performance of the top 10M websites, and maybe more. This is a lesson that every CMS and framework should heed. The more we can set up smart defaults for performance at the framework level, the better we can make the web while also make developers' jobs easier.

Reviewer: > Finally,

to

> Fourth,

or

> Finally, fourth,

Just to be consistent with the format you started.

> we saw that WordPress can impact the performance of the top 10M websites, and maybe more.

to

> we saw that WordPress impacted the performance of the top 10M websites (maybe more).

and

> The more we can set up smart defaults for performance at the framework level, the better we can make the web while also make developers' jobs easier.

to

> When we set smarter defaults at the framework level, the better the performance will be on the Web.

Author: The last rephrase removed the concept of making developer's jobs easier.

Reviewer: I guess my feeling here is that I don't think that's your goal.  I think your goal is to make performance better and more accurate.  The by-product is the developer's job becoming easier, in this very narrow aspect.  I challenge the idea because I don't think you know if it would be easier (and what other complications this would mean if defaults are adopted).  I.e., would having defaults mean dev need to now build dashboards to view this in some way?  Will they have to pipeline it and store it and build a system to look at this?  Are additional monitoring steps needed by ops or dev to track this information (where before maybe it was on an individual basis and not a standard everyone follows)?  I think the developer job being easier is speculation.  I also, honestly, just couldn't fit it in and make it not sound like an extra by-product, structurally.  Probably because I was biased that it wasn't your end-game.

Author: No, developer experience is very important. Performance is too difficult which is why it is so bad. It's too much to learn combined with everything else on the front end. So the more we can do it at the framework level, the better it will be.

Reviewer: OK.  Then it deserves its own sentence (if not, its own paragraph), not as an added thought at the end of the initial idea of improving the web experience for users.  For example:

> When we set smarter defaults on the framework level, the performance will improve on the Web.  This is because developers would no longer be individually solving the complex problem of capturing performance data.  Instead, they can focus on monitoring and responding to the consistent and correctly captured data provided by the framework.

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: Performance
description: Performance chapter of the 2021 Web Almanac covering Core Web Vitals (Largest Contentful Paint, Cumulative Layout Shift, First Input Delay) as well as First Contentful Paint and Time to First Byte
authors: [siakaramalegos]
reviewers: [rviscomi, kevinfarrugia, estelle, ziemek-bucko, jzyang, fili, samarpanda, edmondwwchan]
analysts: [siakaramalegos, rviscomi, Nithanaroy]

> editors: [jzyang]

Reviewer 2: ```suggestion
editors: [jzyang, tunetheweb]
```

translators: []
results: https://docs.google.com/spreadsheets/d/13xhPx6o2Nowz_3b3_5ojiF_mY3Lhs25auBKM6eqGZmo/
siakaramalegos_bio: Sia Karamalegos is a web developer, international conference speaker, and writer. She is a Google Developer Expert in Web Technologies, a Cloudinary Media Developer Expert, a Stripe Community Expert, and co-organizes the Eleventy Meetup. Check out her writing, speaking, and newsletter on <a hreflang="en" href="https://sia.codes/">sia.codes</a> or find her on <a hreflang="en" href="https://twitter.com/thegreengreek">Twitter</a>.
featured_quote: TODO
featured_stat_1: TODO
featured_stat_label_1: TODO
featured_stat_2: TODO
featured_stat_label_2: TODO
featured_stat_3: TODO
featured_stat_label_3: TODO

> unedited: true

Reviewer 2: I'm happy to remove this with @jzyang 's great work and my run-through there.

```suggestion
```

---

## Introduction

Performance is important for user experience. Slow-to-load and slow-to-respond websites frustrate users and cause lost conversions. This is the first year that the <a hreflang="en" href="https://web.dev/vitals/">Core Web Vitals</a> have contributed to Google search rankings. As such, we've seen greater interest in improving website performance which is great news for users.

> *What are our top takeaways from this year's report?* First, we still have a long way to go in providing a good user experience. For example, faster networks and devices have not yet reached the point where we can ignore how much JavaScript we deliver to a site; and, we may never get there. Second, sometimes we misuse new features for performance, resulting in poorer performance. Third, we need better metrics for measuring interactivity, and those are on the way. And fourth, CMS- and framework-level work on performance can significantly impact user experience for the top 10M websites.

Author: The comma is needed or it's an incomplete sentence

> *What's new this year?* We're excited to share performance data by traffic ranking for the first time. We also have all the core performance metrics from previous years. Finally, we added a deeper dive into the Largest Contentful Paint (LCP) element.

Author: I could go either way on that.

### Notes on Methodology

> One thing that makes the performance chapter different from the others is that we rely heavily on the <a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a> (CrUX) for our analyses. Why? If our number one priority is user experience, then the best way to measure performance is with real user data (real user metrics, or RUM for short).

Author: I think you're reading too much into this. Don't let the perfect be the enemy of the good.

> The Chrome User Experience Report provides user experience metrics for how real-world Chrome users experience popular destinations on the web. -<a hreflang="en" href="https://developers.google.com/web/tools/chrome-user-experience-report">Chrome User Experience Report</a>

CrUX data only provides high-level field/RUM metrics and only for the Chrome browser. Additionally, CrUX reports data by origin, or website, instead of by page.

> We supplement our CrUX RUM data with lab data from WebPageTest in HTTP Archive. WebPageTest includes very detailed information about each page, including the full Lighthouse report. Note that WebPageTest measures performance in <a hreflang="en" href="https://almanac.httparchive.org/en/2021/methodology#webpagetest">California and Oregon</a>. The performance data in CrUX is global since it represents real user page loads.

Author: @rviscomi told me it was California and Oregon so that is why I put that. The reason why I want to put it is because I've seen folks in other parts of the world complain that their sites are getting unfairly judged due to testing location.

> We supplement our CrUX RUM data with lab data from WebPageTest in HTTP Archive. WebPageTest includes very detailed information about each page, including the full Lighthouse report. Note that WebPageTest measures performance in <a hreflang="en" href="https://almanac.httparchive.org/en/2021/methodology#webpagetest">California and Oregon</a>. The performance data in CrUX is global since it represents real user page loads.

Author: No on "therefore"

> We supplement our CrUX RUM data with lab data from WebPageTest in HTTP Archive. WebPageTest includes very detailed information about each page, including the full Lighthouse report. Note that WebPageTest measures performance in <a hreflang="en" href="https://almanac.httparchive.org/en/2021/methodology#webpagetest">California and Oregon</a>. The performance data in CrUX is global since it represents real user page loads.

Author: changed to U.S.

When comparing performance year-over-year, keep in mind that:

- The Cumulative Layout Shift (CLS) calculation has changed since 2020.
- The First Contentful Paint (FCP) thresholds ("good", "needs improvement", and "poor") have changed since 2020.
- Last year's report was based on August 2020 data, and this year's report was based on the July 2021 run.

> Read the full [methodology](./methodology) for the Web Almanac to learn more.

Author: I think removing the last item as a bullet makes it easy to skip over. So will make the other changes but put it back as a bullet.

> Read the full [methodology](./methodology) for the Web Almanac to learn more.

Reviewer 2: Was trying to separate CWV changes (and link for that), from Web Almanac changes (and link for that).

I'm also not convinced we even need to mention the July -> August thing. I think it adds unnecessary confusion to this chapter and is best handled in Methodology page. One month doesn't make a lot of difference and it was done for project reasons rather than any particular measurement reasons.

## High-Level Performance: Core Web Vitals

> Before we dive into the individual metrics, let's take a look at combined performance for <a hreflang="en" href="https://web.dev/vitals/">Core Web Vitals</a> (CWV). Core Web Vitals (LCP, CLS, FID) are a set of performance metrics focused on user experience. They focus on loading, interactivity, and visual stability.

Author: lol same but 🤷

Web performance is notorious for an alphabet soup of metrics, but the community is coalescing on this framework.

This section focuses on websites that reached the "good" threshold on all three CWV metrics to understand how the web is performing at a high level. In the Analysis by Metric section, we'll cover the same charts by each metric in detail, plus more metrics not in the CWV.

### By Device

{{ figure_markup(
  image="performance-1-good-cwv-by-device.png",
  caption="Good Core Web Vitals by Device from 2020 to 2021",
  description="Bar chart showing the percent of origins with good Core Web Vitals in 2020 and 2021. In 2020, 34% of desktop websites were good and 24% of mobile websites. In 2021, 41% of desktop websites were good, and 29% of mobile ones.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=116267418&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

<!-- TODO: note style -->

> <p class="note"><strong>Note:</strong> As the CLS calculation changed since last year, this is not an apples-to-apples comparison.</p>

Author: The other calculations did not change, only the thresholds.

> <p class="note"><strong>Note:</strong> As the CLS calculation changed since last year, this is not an apples-to-apples comparison.</p>

Reviewer 2: Well not strictly true. For example, LCP has had several improvements to what it considers the LCP elements - which I'd argue are calculation changes: https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/speed/metrics_changelog/lcp.md

![image](https://user-images.githubusercontent.com/10931297/141792537-11600a76-e2de-4c28-9d51-a7eeaeb54597.png)

> Core Web Vitals for websites in the Chrome User Experience Report improved year-over-year. But, a good part of this improvement could be due to a change in the CLS calculation, not necessarily to a performance improvement in CLS. The resulting CLS "improvement" was 8 points on desktop (2 for mobile). LCP improved by 7 points for desktop (2 for mobile). FID was already at 100% for desktop for both years and improved by 10 points on mobile.

Author: Comma is needed to make it a complete sentence.

### By Effective Connection Type

The data by connection type in CrUX can be difficult to understand. It is not based on traffic. If a website has any experiences in a connection type, then it increases the denominator for that connection type. If the experiences were good for that website in that connection type, then it increases the numerator. Said another way, for all the websites which experienced page loads at 4G speed, 36% of those websites had good CWV:

{{ figure_markup(
  image="performance-2-good-cwv-by-ect.png",
  caption="Good CWV performance by effective connection type",
  description="Bar chart showing the percent of origins in a connection type with good Core Web Vitals. The offline category had 44% of websites with good CWV, 0% for slow 2G and 2G, 5% for 3G, and 36% for 4G.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=150765595&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_eff_connection_type.sql"
  )
}}

> Faster connections correlated with better Core Web Vitals performance. Offline performance was better presumably because of service worker caching in progressive web apps. Yet, the number of origins in the offline effective connection type category is negligible at 2,634 total (0.02%).

Author: The case I'm describing here is only for offline sites with service workers, not all service worker websites. I think we would need to include the data to make that point and I saw lots of caveats with it so not sure this is the year to include it.

> Faster connections correlated with better Core Web Vitals performance. Offline performance was better presumably because of service worker caching in progressive web apps. Yet, the number of origins in the offline effective connection type category is negligible at 2,634 total (0.02%).

Author: No need to state the obvious :)

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1955368947&format=interactive",
  width=600,
  height=792,
  sheets_gid="730149797",
  sql_file="web_vitals_by_country.sql"
  )
}}

The US, Japan, Brazil, and Western Europe are over-represented in the CrUX dataset origins. Many large countries like China, India, Nigeria, and Pakistan are underrepresented.

Remember that CrUX data is only gathered in Chrome. The percent of origins by country does not align with relative population sizes. Reasons may include differences in browser share, device share, level of access, and level of use. Also, the origins accessed in that country may not mirror the global top origins included in the dataset. Keep these caveats in mind when evaluating regional-level differences and context for all CrUX analyses.

### By Rank

> This year for the first time, we have ranking data! ​​CrUX determines ranking by the number of page views per website measured in Chrome. In the charts, the categories are additive. The top 10,000 sites include the top 1,000 sites, and so forth. The "all" category is sometimes referred to as the top 10,000,000, but it does not have a full 10 million websites in it. See the [methodology](./methodology#chrome-ux-report) for more details.

Author: But that would affect the other chapter too, right?

> This year for the first time, we have ranking data! ​​CrUX determines ranking by the number of page views per website measured in Chrome. In the charts, the categories are additive. The top 10,000 sites include the top 1,000 sites, and so forth. The "all" category is sometimes referred to as the top 10,000,000, but it does not have a full 10 million websites in it. See the [methodology](./methodology#chrome-ux-report) for more details.

Reviewer 2: Yup. Need to be consistent here, so keeping an eye out for this on all chapters.

  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

For rank, TTFB was faster for higher-ranking sites. One reason could be that most of these are larger companies with more resources to prioritize performance. They may focus on improving server-side performance and delivering assets through edge CDNs. Another reason could be selection bias - the top origins might be accessed more in regions with closer servers, i.e., lower latency.

<!-- TODO: validate CMS chapter link works -->
One more possibility has to do with CMS adoption. The [CMS Chapter](./cms) shows CMS adoption by rank.

{{ figure_markup(
  image="cms-adoption-by-rank.png",
  caption="CMS adoption by rank",
  description="Bar chart showing percent of websites using a CMS for each ranking category. On mobile, the numbers are 7% of websites in the top 1,000, 15% in the top 10,000, 20% in the top 100,000, 29% in the top 1,000,000, and 42% in the top 10,000,000.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRzYkQYaBTpD5FKloDXsdb32Y2pjdxFZq_LoekECgTy53F-dRATm9wFFJ3dDVDS8_cGn0h2mKDBjgdM/pubchart?oid=409766380&format=interactive",

>   sheets_gid="730149797",

Reviewer 2: Needs to be the full URL as not in this chapter's sheet.

Also looks like they changed the sheet id as not the same?

```suggestion
  sheets_gid="https://docs.google.com/spreadsheets/d/1gAJh4VcSEU6-6aQxBiTFl-7cbyHukfdhg-Iaq9y5pnc/#gid=158167539",
```

If there's any other borrowed sheets then similar will need to be done.

  sql_file="../cms/cms_adoption_by_rank.sql"
  )
}}

42% of pages (mobile) in the "all" group used a CMS whereas the top 1,000 sites only had 7% adoption.

> Then, if we look at the top 5 CMSs by rank, we see that WordPress has the highest adoption at for 33.6% of "all" pages:

Author: So the labels are confusing because they don't match the color of the bar they are referring to. I'll look for it in the original spreadsheet and make some more edits.

{{ figure_markup(
  image="top-cmss-by-rank.png",
  caption="Top 5 CMSs by rank",
  description="Bar chart showing the top CMSs percent of share of each ranking category. WordPress had 3.1% of the top 1,000, 8.6% of the top 10,000, 13.2% of the top 100,000, 21.1% of the top 1,000,000, and 33.6% of the top 10,000,000. The remaining CMSs had less than 3% share of all rankings.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRzYkQYaBTpD5FKloDXsdb32Y2pjdxFZq_LoekECgTy53F-dRATm9wFFJ3dDVDS8_cGn0h2mKDBjgdM/pubchart?oid=2087254046&format=interactive",

>   sheets_gid="730149797",

Reviewer 2: Needs to be the full URL as not in this chapter's sheet:

```suggestion
  sheets_gid="https://docs.google.com/spreadsheets/d/1gAJh4VcSEU6-6aQxBiTFl-7cbyHukfdhg-Iaq9y5pnc/#gid=670045665",
```

If there's any other borrowed sheets then similar will need to be done.

  )
}}

Only 5% of origins on WordPress experienced good TTFB in July 2021. Considering WordPress's large share of the top 10M sites, its poor TTFB could be a contributor to the TTFB degradation by rank.

### First Contentful Paint (FCP)

<a hreflang="en" href="https://web.dev/fcp/">First Contentful Paint (FCP)</a> measures the time from when a load first begins until the browser first renders any contentful part of the page (e.g, text, images, etc.).

{{ figure_markup(
  image="performance-FCP-by-device.png",
  caption="FCP performance by device",
  description="Stacked bar chart showing FCP performance for desktop and mobile devices. For desktop devices, 60% of websites had good (< 1.8s), 27% needs improvement, and 13% poor (>= 3.0s) FCP performance. For mobile, 38% were good, 38% needs improvement, and 24% poor.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1509111466&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

> FCP was faster on desktop than mobile, likely due to both faster average network speeds and faster processors. Only 38% of origins had "good" FCP on mobile. Render-blocking resources such as synchronous JavaScript can be a common culprit. Because TTFB is the first part of FCP, poor TTFB will make it difficult to achieve a good FCP.

Author: Good point. I used quotes when I was referring to the category group directly rather than just describing the results. I could try to make it more consistent or just remove the quotes.

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1080361066&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

For LCP, the differences in performance by rank were closer than FCP. Also, a higher proportion of origins in the top 1,000 had poor LCP. On WordPress, <a hreflang="en" href="https://datastudio.google.com/s/kvq1oJ60jaQ">28% of origins experienced good LCP</a>. This is an opportunity to improve user experience as poor LCP is usually caused by a handful of problems.

### The LCP Element

Let's take a deeper dive into the LCP element.

{{ figure_markup(
  image="performance-top-15-lcp-nodes.png",
  caption="Top 15 LCP HTML element nodes",

>   description="Bar chart showing the percent share of pages for each node type for the LCP element- data listed by desktop then mobile: IMG 47% 42%, DIV 28% 27%, P 6% 10%, H1 3% 5%, SECTION 3% 2%, undetected 2% 2%, H2 2% 2%, A 1% 2%, SPAN 1% 1%, HEADER 1% 1%, LI 1% 1%, RS-SBG 1% 1%, H3 0% 1%, TD 1% 1%, VIDEO 0% 0%",

Author: I don't think the code formatting is needed since it's not actually code - just a listing of the node names. It's probably a web component.

>   description="Bar chart showing the percent share of pages for each node type for the LCP element- data listed by desktop then mobile: IMG 47% 42%, DIV 28% 27%, P 6% 10%, H1 3% 5%, SECTION 3% 2%, undetected 2% 2%, H2 2% 2%, A 1% 2%, SPAN 1% 1%, HEADER 1% 1%, LI 1% 1%, RS-SBG 1% 1%, H3 0% 1%, TD 1% 1%, VIDEO 0% 0%",

Reviewer 2: Part of the reason for code formatting is the readability as otherwise the numbers in the headings bleed into the percentages:

![image](https://user-images.githubusercontent.com/10931297/141794622-ae102a4a-2394-4ca8-8b90-a1d6d8203ad7.png)

Here's what it looks like with the code formatting:

![image](https://user-images.githubusercontent.com/10931297/141794843-9bef87f3-743a-4778-9e0d-2f3001b40ff5.png)

(still don't like the lighter code formatting on the dark grey background - but that's a completely different issue and not your problem!)

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1549429021&format=interactive",
  width=600,
  height=909,
  sheets_gid="730149797",
  sql_file="lcp_element_data.sql"
  )
}}

IMG, DIV, P, and H1 made up 83% of all LCP nodes (on mobile). This doesn't tell us if the content was an image or text, as background images can be applied with CSS.

{{ figure_markup(
  image="performance-lcp-with-image.png",
  caption="LCP elements with images, by device",
  description="Bar chart showing 79% of desktop pages have an LCP element with an image and 71% of mobile pages",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=992354142&format=interactive",
  sheets_gid="730149797",
  sql_file="lcp_element_data.sql"
  )
}}

> We can see that 71-79% of pages had an LCP element that was an image, regardless of HTML node. Furthermore, desktop devices had a higher rate of LCPs as images. This could be due to less real estate on smaller screens pushing images out of the viewport resulting in heading text being the largest element.

Author: `COUNTIF(elementUrl != "") AS haveImages,`

We don't explain the other queries so that's my only hesitation in explaining it here, but I could go either way.

In both cases, images comprised the majority of LCP elements. This warrants a deeper dive into how those images are loading.

{{ figure_markup(
  image="performance-lcp-antipatterns.png",
  caption="LCP elements with performance anti-patterns",

>   description="Bar chart showing 9.3% of pages on mobile use native lazy load on the LCP element, 16.5% probably lazy load, and 0.4% use async decode.",

Reviewer 2: "native lazy load" or "native lazy-load" or "native lazy-loading".

Seems to be a lot of hyphenated and unhyphenated usage. Are they all correct?

Author: Depends on usage. I think here it does not need a hyphen.

>   chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=130632749&format=interactive",

Reviewer 2: I think there's plenty of room for the green label in this chart.

Author: I defaulted to "don't use it unless absolutely necessary since it's terrible for accessibility"

  sheets_gid="730149797",
  sql_file="lcp_element_data.sql"
  )
}}

For user experience, we want LCP elements to load as fast as possible. User experience is why LCP was selected as one of the Core Web Vitals. We do not want it to be lazy-loaded nor decoded asynchronously as that further delays the render. However, we can see that 9.3% of pages used the native loading=lazy flag on the LCP `<img>` element.

Not all browsers support native lazy loading. Popular lazy loading polyfills detect a "lazyload" class on an image element. Thus, we can identify more possibly lazy-loaded images by adding images with a "lazyload" class to the total. The percent of sites probably lazy loading their LCP `<img>` element jumps up to 16.5% on mobile.

> Lazy loading your LCP element will result in worse performance. *Don't do it!* WordPress was an early adopter of native lazy loading. The early method was a naive solution applying lazy loading to all images, and the results showed a <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">negative performance correlation</a>. They were able to use this data to implement a more nuanced approach for better performance.

Reviewer 2: Note it was only JUST committed: https://github.com/WordPress/wordpress-develop/pull/1580#issuecomment-963704254. Not sure if you want to link to that issue?

```suggestion
Lazy loading your LCP element will result in worse performance. *Don't do it!* WordPress was an early adopter of native lazy loading. The early method was a naive solution applying lazy loading to all images, and the results showed a <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">negative performance correlation</a>. They were able to use this data to implement a <a hreflang="en" href="https://github.com/WordPress/wordpress-develop/pull/1580">more nuanced approach for better performance</a>.
```

Author: Good suggestion!

> Luckily, only 0.4% of sites used the async decode directive for their LCP image. The negative impact of asynchronous decode is not as high as lazy loading.

Reviewer 2: Why is this bad?

I really don't get this attribute, but here was my latest understanding which suggests it SHOULD be used: https://twitter.com/tunetheweb/status/1394683568874475522?s=20

Author: I'm re-reading that chapter in Addy's book, and I still think it might be the wrong decision for the LCP image though I have no data to back it up. Would it be possible to get Addy's input? I understand why you might want it for other images I just have my doubts on the LCP element. @addyosmani

{{ figure_markup(
  content="354",
  caption="Websites attempted to use native lazy-loading on LCP elements that are not images or iframes",
  classes="big-number",
  sheets_gid="1423728540",
  sql_file="lcp_element_data.sql"
  )
}}

> Interestingly, 354 origins on desktop attempted to use native lazy-loading on HTML elements that do not support the loading attribute (e.g., `<div>`). The loading attribute is only supported on `<img>` and  sometimes `<iframe>` elements (see <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">caniuse</a>).

Reviewer 2: "sometimes" seems a little weak.

```suggestion
Interestingly, 354 origins on desktop attempted to use native lazy-loading on HTML elements that do not support the loading attribute (e.g., `<div>`). The loading attribute is only supported on `<img>` and, for fewer browsers, `<iframe>` elements (see <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">caniuse</a>).
```

Author: I get your point but I think I'm going to use a different word so keeping this open so I remember to change it.

### Cumulative Layout Shift (CLS)

{{ figure_markup(
  image="performance-CLS-by-device.png",
  caption="CLS performance by device",
  description="Stacked bar chart showing CLS performance for desktop and mobile devices. For desktop devices, 62% of websites had good (< 0.1), 23% needs improvement, and 15% poor (>= 0.25) CLS performance. For mobile, 62% were good, 21% needs improvement, and 17% poor.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=160840238&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_device.sql"
  )
}}

> <a hreflang="en" href="https://web.dev/cls/">Cumulative Layout Shift (CLS)</a> is characterized by how much shift a user experiences, not how long it takes to visually see something like FCP and LCP. As such, performance by device was fairly equivalent.

Reviewer 2: "how much shift a user experiences" is a little non-descriptive.

```suggestion
<a hreflang="en" href="https://web.dev/cls/">Cumulative Layout Shift (CLS)</a> is characterized by how much content moves around or "shifts" while viewing a web page, not how long it takes to visually see something like FCP and LCP. As such, performance by device was fairly equivalent.
```

Author: That seems long winded and more difficult to read, so I might change it to something else. Make sure you run complex sentences through Hemingway.app

Unlike the other metrics, FID was worse for offline websites than any other connection category. This could be due to the more complex nature of many websites with service workers. Having a service worker does not eliminate the impact of client-side JavaScript running on the main thread.

{{ figure_markup(
  image="performance-FID-by-rank.png",
  caption="FID performance by rank",
  description="Stacked bar chart showing FID performance for each ranking group. For all categories, 93-94% had good, 5-7% needs improvement, and 0% poor performance.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=1539782601&format=interactive",
  sheets_gid="730149797",
  sql_file="web_vitals_by_rank.sql"
  )
}}

FID performance by rank was flat.

For all FID metrics, we see very large bars in the "good" category which makes it less effective unless we've truly hit peak performance. The good news is the <a hreflang="en" href="https://web.dev/better-responsiveness-metric/">Chrome team is evaluating this now</a> and would like your feedback.

> If your site's performance is not in the "good" category, then you definitely have a performance problem. A common culprit for FID issues is too much long-running JavaScript. Keep your bundle sizes small and pay attention to third-party scripts.

Reviewer 2: ```suggestion
If your site's performance is not in the "good" category, then you definitely have a performance problem. A common culprit for FID issues is too much long-running JavaScript. Keep your bundle sizes small and pay attention to [third-party](./third-party) scripts.
```

### Total Blocking Time (TBT)

> > The Total Blocking Time (TBT) metric measures the total amount of time between First Contentful Paint (FCP) and Time to Interactive (TTI) where the main thread was blocked for long enough to prevent input responsiveness. --<a hreflang="en" href="https://web.dev/tbt/">Web.dev</a> ]

Reviewer 2: ```suggestion
<figure>
<blockquote>The Total Blocking Time (TBT) metric measures the total amount of time between First Contentful Paint (FCP) and Time to Interactive (TTI) where the main thread was blocked for long enough to prevent input responsiveness.</blockquote>
<figcaption>— <cite><a hreflang="en" href="https://web.dev/tbt/">Web.dev</a></cite></figcaption>
</figure>
```

<a hreflang="en" href="https://web.dev/tbt/">Total Blocking Time (TBT)</a> is a lab-based metric that helps us debug potential interactivity issues. FID is a field-based metric, and TBT is its lab-based analog. Currently, when evaluating client websites, I reach for total blocking time TBT as another indicator of possible performance issues due to JavaScript.

Unfortunately, TBT is not measured in the Chrome User Experience Report. But, we can still get an idea of what's going on using the HTTP Archive Lighthouse data (only collected for mobile):

{{ figure_markup(
  image="performance-tbt.png",
  caption="Lighthouse TBT scores",
  description="Bar chart showing 31% of mobile pages with good TBT (<= 200ms), 11% with needs improvement, and 58% with poor (> 600ms) TBT.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQj6-t49BNV_5W-w83AABkUoo6klfyUyIz13yKShLAzK8qGs5lJ9TKcggIEp6JgxikVF-UJBAHpsrNl/pubchart?oid=2135422883&format=interactive",
  sheets_gid="730149797",
  sql_file="tbt.sql"
  )
}}

> <p class="note"><strong>Note:</strong> The groups in the chart are based off of the Lighthouse score for TBT (e.g., >= 0.9 results in "good"). Due to rounding of the score, some TBT values slightly above 200ms get categorized as "good" (and similarly at the 600ms threshold).</p>

Reviewer 2: Is this necessary? Can't see any of those oddities in the graph so why bother confusing the user?

```suggestion
```

Author: True, they probably would not notice. I just felt dirty having the results be off from what we say they are but that's probably the case for all of them lol.

Remember that the data is a single, throttled-CPU Lighthouse run through WebPageTest and does not reflect real user experiences. Yet, potential interactivity looked much worse when looking at TBT versus FID. The "real" evaluation of your interactivity is probably somewhere between. Thus, if your FID is "good", take a look at TBT in case you're missing some poor user experiences that FID can't catch yet. The same issues that cause poor FID also cause poor TBT.

{{ figure_markup(
  content="67 seconds",
  caption="Longest TBT",
  classes="medium-number",
  sheets_gid="1423728540",
  sql_file="lcp_element_data.sql"
  )
}}

## Conclusion

> We still have a long way to go to provide great user experience, but we can take steps to improve it.

Reviewer 2: Seems a little negative, given that we've improved in last year according to the stats above (though appreciate some of this will be due to metric redefinition - but that's cause they needed redefining).

From the [Author's Guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide):

> **The glass is half-full.** Sometimes we're drawn to shine a light on the fires burning in the tail of the web, but let's not forget about all the good stuff done by folks who are making active efforts to improve their UX. For example, we should highlight the [positive adoption rate of HTTPS](https://httparchive.org/reports/state-of-the-web#pctHttps) and not just dwell on the 20% of websites without it.

Author: I'm trying to balance that as I don't want people to take away that the job is done and they can stop worrying about it which is what I fear will happen if I only say it's positive. Does that make sense?

Author: Changed it to:

> Performance improved since 2020. Though we still have a long way to go to provide great user experience, we can take steps to improve it.

First, you cannot improve performance unless you can measure it. A good first step here is to measure your site using real user devices and to set up real-user monitoring (RUM). You can get a flavor of how your site performs with Chrome users with the <a hreflang="en" href="https://rviscomi.github.io/crux-dash-launcher/">CrUX dashboard launcher</a> (if your site is in the dataset). You should set up a RUM solution that measures across multiple browsers. You can build this yourself or use one of many analytics vendors' solutions.

Second, as new features in HTML, CSS, and JavaScript are released, make sure you understand them before implementing them. Use A/B testing to verify that adopting a new strategy results in improved performance. For example, don't lazy-load images above the fold. If you have a RUM tool implemented, you can better detect when your changes accidentally cause regressions.

> Third, continue to optimize for both FID (field/real-user data) and TBT (lab data). Take a look at the <a hreflang="en" href="https://web.dev/responsiveness/">proposal</a> for a new responsiveness metric and participate by providing feedback. A new <a hreflang="en" href="https://web.dev/smoothness/">animation smoothness metric</a> is also being proposed. In our quest for a faster web, change is inevitable and for the better. As we continue to optimize, you're participation is key.

Author: lol oops

## Introduction

> Performance is important for the user experience. Slow-to-load and slow-to-respond websites frustrate users and cause lost conversions. This is the first year that <a hreflang="en" href="https://web.dev/vitals/">Core Web Vitals</a> have contributed to Google search rankings. As such, we've seen greater interest in improving website performance which is great news for users.

Author: I think "the user experience" sounds awkward. I don't think it's used that way in web dev conversations usually.

Reviewer: I put "the user experience" because that's how it's talked about in the UI/UX circles.

Reviewer: Let me clarify that most of it is not "talked about" but written.

Reviewer: Ex: mostly in the titles though: https://www.usability.gov/what-and-why/user-experience.html

Author: https://twitter.com/TheGreenGreek/status/1459882273415835648

> <i>What are our top takeaways from this year's report?</i> First, we still have a long way to go in providing a good user experience. For example, faster networks and devices have not yet reached the point where we can ignore how much JavaScript we deliver to a site; and, we may never get there. Second, sometimes we misuse new features for performance, resulting in poorer performance. Third, we need better metrics for measuring interactivity, and those are on the way. And fourth, CMS- and framework-level work on performance can significantly impact user experience for the top 10M websites.

Author: `<i>` isn't recommended HTML anymore. In markdown you can wrap in a single asterisk or you can use the `<em>` tag.

Reviewer: Sorry about that.  I just wanted to emphasize the difference.  Feel free to make the change based on what you think is best, if you agree with that format of emphasis.

> <i>What's new this year?</i> We're excited to share performance data by traffic ranking for the first time. We also have all the core performance metrics from previous years. Finally, we added a deeper dive into the Largest Contentful Paint (LCP) element.

Author: Same as above.

Reviewer: Change it how you like.  I just wanted to clarify the clear divide between your question and your answer.
